## 1. Answer the questions

### 1. Derive an analytical solution to the regression problem. Use a vector form of the equation.

Аналитическое решение находят путем минимизации функции потерь MSE (Mean Squared Error).

#### 1.1. Функция потерь в векторной форме (MSE)

Функция потерь $J(\mathbf{w})$ для вектора весов $\mathbf{w}$ имеет вид:

$$
J(\mathbf{w}) = \frac{1}{2m} ||\mathbf{X}\mathbf{w} - \mathbf{y}||^2
$$

Где:
* $\mathbf{w}$ — вектор весов (параметров), который мы ищем.
* $\mathbf{X}$ — матрица признаков (с добавленным столбцом единиц для смещения/intercept).
* $\mathbf{y}$ — вектор целевых значений.
* $m$ — количество обучающих примеров.

#### 1.2. Вывод нормального уравнения

Чтобы найти минимум, необходимо взять частную производную функции потерь по вектору весов $\mathbf{w}$ и приравнять её к нулю ($\nabla_{\mathbf{w}} J(\mathbf{w}) = 0$).

1.  Раскрываем норму и переписываем функцию потерь:
    $$
    J(\mathbf{w}) = \frac{1}{2m} (\mathbf{X}\mathbf{w} - \mathbf{y})^T (\mathbf{X}\mathbf{w} - \mathbf{y})
    $$

2.  Берём градиент:
    $$
    \nabla_{\mathbf{w}} J(\mathbf{w}) = \frac{1}{m} \mathbf{X}^T (\mathbf{X}\mathbf{w} - \mathbf{y})
    $$

3.  Приравниваем градиент к нулю:
    $$
    \mathbf{X}^T (\mathbf{X}\mathbf{w} - \mathbf{y}) = 0
    $$

4.  Выражаем вектор весов $\mathbf{w}$:
    $$
    \mathbf{X}^T \mathbf{X} \mathbf{w} - \mathbf{X}^T \mathbf{y} = 0
    $$
    $$
    \mathbf{X}^T \mathbf{X} \mathbf{w} = \mathbf{X}^T \mathbf{y}
    $$
    $$
    \mathbf{w} = (\mathbf{X}^T \mathbf{X})^{-1} \mathbf{X}^T \mathbf{y}
    $$

#### 1.3. Аналитическое решение

Вектор весов $\mathbf{w}$, который минимизирует MSE, определяется формулой:
$$\mathbf{w} = (\mathbf{X}^T \mathbf{X})^{-1} \mathbf{X}^T \mathbf{y}$$

Это и есть **нормальное уравнение**.

### 2. What changes in the solution when L1 and L2 regularizations are added to the loss function.

Регуляризация добавляет к основной функции потерь MSE член, штрафующий большие значения весов $\mathbf{w}$.

#### 2.1. L2 Регуляризация (Ridge Regression)

* **Функция потерь:** $J_{\text{Ridge}}(\mathbf{w}) = J(\mathbf{w}) + \alpha ||\mathbf{w}||_2^2$
    $$
    J_{\text{Ridge}}(\mathbf{w}) = \frac{1}{2m} ||\mathbf{X}\mathbf{w} - \mathbf{y}||^2 + \alpha \sum_{j=1}^{n} w_j^2
    $$
    Где $\alpha \ge 0$ — коэффициент регуляризации.

* **Аналитическое решение (Нормальное уравнение для Ridge):**
    Добавление L2-штрафа делает матрицу $(\mathbf{X}^T \mathbf{X})$ всегда обратимой (даже при мультиколлинеарности), стабилизируя решение.
    $$\mathbf{w}_{\text{Ridge}} = (\mathbf{X}^T \mathbf{X} + \alpha \mathbf{I})^{-1} \mathbf{X}^T \mathbf{y}$$
    Где $\mathbf{I}$ — единичная матрица, и $\mathbf{I}$ имеет размерность $n \times n$ (где $n$ — количество признаков).

#### 2.2. L1 Регуляризация (Lasso Regression)

* **Функция потерь:** $J_{\text{Lasso}}(\mathbf{w}) = J(\mathbf{w}) + \alpha ||\mathbf{w}||_1$
    $$
    J_{\text{Lasso}}(\mathbf{w}) = \frac{1}{2m} ||\mathbf{X}\mathbf{w} - \mathbf{y}||^2 + \alpha \sum_{j=1}^{n} |w_j|
    $$

* **Аналитическое решение (Lasso):**
    Из-за наличия модуля ($|w_j|$) функция потерь для Lasso **недифференцируема в нуле**, то есть у неё **не существует замкнутого аналитического решения** (аналогичного нормальному уравнению).

    Поэтому Lasso-регрессия решается **итерационными методами**, такими как:
    1.  **Координатный спуск (Coordinate Descent).**
    2.  **Субградиентный спуск (Subgradient Descent)** или, чаще, модификации Градиентного спуска (например, FISTA).

### 3. Explain why L1 regularization is often used to select features. Why are there many weights equal to 0 after the model is fit?

L1-регуляризация (Lasso) используется для выбора признаков, поскольку она приводит к тому, что веса **многих неинформативных признаков становятся точно равны нулю**. 

#### 3.1. Геометрическая причина (Контурные линии)

Это свойство можно объяснить, рассмотрев геометрическое представление минимизации. Задача минимизации заключается в поиске точки касания между:
1.  **Эллиптическими контурами** функции потерь MSE.
2.  **Областью ограничений** регуляризатора.

* **L2 (Ridge):** Область ограничений — это **круг** (или сфера). Точка касания с эллипсом MSE почти всегда происходит **не на осях**, а где-то в пространстве. Следовательно, веса $w_j$ становятся **близки к нулю**, но редко *равны* нулю.

* **L1 (Lasso):** Область ограничений — это **ромб** (или многогранник с "острыми углами" на осях). Когда контурная линия MSE касается ромба, она **очень часто касается его в одном из этих углов**. Точка касания, находящаяся на оси, соответствует ситуации, когда вес, связанный с этой осью, **обнуляется** (становится $w_j = 0$).

#### 3.2. Почему веса становятся 0

Математически, штраф $L1$ ($\sum |w_j|$) имеет тенденцию "выталкивать" веса к нулю, если они не вносят существенного вклада в снижение ошибки.

Когда модель обучена (fit), веса, равные 0, означают, что **соответствующий признак был исключен из модели**. Таким образом, L1-регуляризация автоматически выполняет **выбор признаков (Feature Selection)**.

### 4. Explain how you can use the same models (Linear regression, Ridge, etc.) but make it possible to fit nonlinear dependencies.

Модели, такие как Линейная регрессия, Ridge и Lasso, по своей природе являются **линейными по параметрам (весам)**, но они могут моделировать **нелинейные зависимости между признаками и таргетом**.

#### 4.1. Метод: Полиномиальная регрессия (Polynomial Features)

Ключевой метод — **создание нелинейных признаков** из существующих **до** обучения модели. Это достигается с помощью **полиномиальных преобразований** и **взаимодействий признаков (Interaction Terms)**.

**Пример:**
Пусть исходная зависимость нелинейна: $y = w_0 + w_1 x_1^2 + w_2 x_2 + \epsilon$

1.  **Трансформация признаков:** Мы создаём новый признак $x_{new} = x_1^2$.
2.  **Новая модель:** Модель становится линейной по новым признакам: $y = w_0 + w_1 x_{new} + w_2 x_2 + \epsilon$.

Этот процесс называется **Полиномиальная регрессия**. 

#### 4.2. Алгоритм

1.  **Создание полиномиальных признаков:** Используем `sklearn.preprocessing.PolynomialFeatures` для создания признаков высокой степени ($x^2, x^3, \dots$) и признаков взаимодействия ($x_1 x_2, x_1^2 x_2, \dots$) из исходной матрицы $\mathbf{X}$.
2.  **Нормализация:** Обязательно применяем **StandardScaler** к новым, расширенным признакам, чтобы стабилизировать обучение.
3.  **Обучение:** Обучаем линейную модель (Linear Regression, Ridge, Lasso) на **трансформированных признаках**.

Таким образом, модель остается "линейной" в пространстве своих параметров $\mathbf{w}$, но благодаря преобразованию входных данных может аппроксимировать сколь угодно сложную **нелинейную кривую** в исходном пространстве признаков.

## 2. Introduction — make all the preprocessing staff from the previous lesson

### 1. Import libraries.

In [1]:
import pandas as pd
import numpy as np
import seaborn as sns
import matplotlib.pyplot as plt
from collections import Counter
import random

from sklearn.linear_model import LinearRegression, Ridge, Lasso, ElasticNet
from sklearn.preprocessing import MinMaxScaler, StandardScaler, PolynomialFeatures

In [2]:
RANDOM_SEED = 21

random.seed(RANDOM_SEED)
np.random.seed(RANDOM_SEED)

### 2. Read Train and Test Parts.

In [3]:
df_train = pd.read_json("data/train.json")
df_train.head(5)

,bathrooms,bedrooms,building_id,created,description,display_address,features,latitude,listing_id,longitude,manager_id,photos,price,street_address,interest_level
4,1.0,1,8579a0b0d54db803821a35a4a615e97a,2016-06-16 05:55:27,Spacious 1 Bedroom 1 Bathroom in Williamsburg!...,145 Borinquen Place,"[Dining Room, Pre-War, Laundry in Building, Di...",40.7108,7170325,-73.9539,a10db4590843d78c784171a107bdacb4,[https://photos.renthop.com/2/7170325_3bb5ac84...,2400,145 Borinquen Place,medium
6,1.0,2,b8e75fc949a6cd8225b455648a951712,2016-06-01 05:44:33,BRAND NEW GUT RENOVATED TRUE 2 BEDROOMFind you...,East 44th,"[Doorman, Elevator, Laundry in Building, Dishw...",40.7513,7092344,-73.9722,955db33477af4f40004820b4aed804a0,[https://photos.renthop.com/2/7092344_7663c19a...,3800,230 East 44th,low
9,1.0,2,cd759a988b8f23924b5a2058d5ab2b49,2016-06-14 15:19:59,**FLEX 2 BEDROOM WITH FULL PRESSURIZED WALL**L...,East 56th Street,"[Doorman, Elevator, Laundry in Building, Laund...",40.7575,7158677,-73.9625,c8b10a317b766204f08e613cef4ce7a0,[https://photos.renthop.com/2/7158677_c897a134...,3495,405 East 56th Street,medium
10,1.5,3,53a5b119ba8f7b61d4e010512e0dfc85,2016-06-24 07:54:24,A Brand New 3 Bedroom 1.5 bath ApartmentEnjoy ...,Metropolitan Avenue,[],40.7145,7211212,-73.9425,5ba989232d0489da1b5f2c45f6688adc,[https://photos.renthop.com/2/7211212_1ed4542e...,3000,792 Metropolitan Avenue,medium
15,1.0,0,bfb9405149bfff42a92980b594c28234,2016-06-28 03:50:23,Over-sized Studio w abundant closets. Availabl...,East 34th Street,"[Doorman, Elevator, Fitness Center, Laundry in...",40.7439,7225292,-73.9743,2c3b41f588fbb5234d8a1e885a436cfa,[https://photos.renthop.com/2/7225292_901f1984...,2795,340 East 34th Street,low


In [4]:
df_test = pd.read_json("data/test.json")
df_test.head(5)

,bathrooms,bedrooms,building_id,created,description,display_address,features,latitude,listing_id,longitude,manager_id,photos,price,street_address
0,1.0,1,79780be1514f645d7e6be99a3de696c5,2016-06-11 05:29:41,Large with awesome terrace--accessible via bed...,Suffolk Street,"[Elevator, Laundry in Building, Laundry in Uni...",40.7185,7142618,-73.9865,b1b1852c416d78d7765d746cb1b8921f,[https://photos.renthop.com/2/7142618_1c45a2c8...,2950,99 Suffolk Street
1,1.0,2,0,2016-06-24 06:36:34,Prime Soho - between Bleecker and Houston - Ne...,Thompson Street,"[Pre-War, Dogs Allowed, Cats Allowed]",40.7278,7210040,-74.0000,d0b5648017832b2427eeb9956d966a14,[https://photos.renthop.com/2/7210040_d824cc71...,2850,176 Thompson Street
2,1.0,0,0,2016-06-17 01:23:39,Spacious studio in Prime Location. Cleanbuildi...,Sullivan Street,"[Pre-War, Dogs Allowed, Cats Allowed]",40.7260,7174566,-74.0026,e6472c7237327dd3903b3d6f6a94515a,[https://photos.renthop.com/2/7174566_ba3a35c5...,2295,115 Sullivan Street
3,1.0,2,f9c826104b91d868e69bd25746448c0c,2016-06-21 05:06:02,For immediate access call Bryan.<br /><br />Bo...,Jones Street,"[Hardwood Floors, Dogs Allowed, Cats Allowed]",40.7321,7191391,-74.0028,41735645e0f8f13993c42894023f8e58,[https://photos.renthop.com/2/7191391_8c2f2d49...,2900,23 Jones Street
5,1.0,1,81062936e12ee5fa6cd2b965698e17d5,2016-06-16 07:24:27,Beautiful TRUE 1 bedroom in a luxury building ...,Exchange Place,"[Roof Deck, Doorman, Elevator, Fitness Center,...",40.7054,7171695,-74.0095,a742cf7dd3b2627d83417bc3a1b3ec96,[https://photos.renthop.com/2/7171695_089ffee2...,3254,20 Exchange Place


## 3. Intro data analysis part 2

In [5]:
def remove_outliers(df, columns):
    mask = pd.Series([True] * len(df), index=df.index)

    for col in columns:
        Q1 = np.percentile(df[col], 1)
        Q99 = np.percentile(df[col], 99)
        print(f"{col}: Q1 = {Q1}, Q99 = {Q99}")
        mask &= (df[col] >= Q1) & (df[col] <= Q99)
    
    return df[mask]

df_train = remove_outliers(df_train, ['price'])
print("\n")
df_test = remove_outliers(df_test, ['price', 'bathrooms'])

price: Q1 = 1475.0, Q99 = 13000.0


price: Q1 = 1495.0, Q99 = 13000.0
bathrooms: Q1 = 1.0, Q99 = 3.0


### 1. Let's generate additional features for better model quality. Consider a column called "Features". It consists of a list of highlights of the current flat.

### 2. Remove unused symbols ([,], ', ", and space) from the column.

In [6]:
def clean_features(series):
  return (
    series.astype(str)
    .str.replace('[', '', regex=False)
    .str.replace(']', '', regex=False)
    .str.replace("'", '', regex=False)
    .str.replace('"', '', regex=False)
    .str.replace(' ', '', regex=False)
  )

df_train['features'] = clean_features(df_train['features'])
df_test['features'] = clean_features(df_test['features'])

In [7]:
df_train['features']

4         DiningRoom,Pre-War,LaundryinBuilding,Dishwashe...
6         Doorman,Elevator,LaundryinBuilding,Dishwasher,...
9         Doorman,Elevator,LaundryinBuilding,LaundryinUn...
10                                                         
15         Doorman,Elevator,FitnessCenter,LaundryinBuilding
                                ...                        
124000                   Elevator,Dishwasher,HardwoodFloors
124002    CommonOutdoorSpace,CatsAllowed,DogsAllowed,NoF...
124004    DiningRoom,Elevator,Pre-War,LaundryinBuilding,...
124008    Pre-War,LaundryinUnit,Dishwasher,NoFee,Outdoor...
124009    DiningRoom,Elevator,LaundryinBuilding,Dishwash...
Name: features, Length: 48379, dtype: object

### 3. Get all values in each list and collect the result in one huge list for the whole dataset. You can use DataFrame.iterrows().

In [8]:
all_features = []
for index, row in df_train.iterrows():
  all_features.extend(row['features'].split(','))

### 4. How many unique values does a result list contain?

In [9]:
print(len(set(all_features)))

1530


### 5. Let's get acquainted with the new library — Collections. With this package you could effectively get quantity statistics about your data.

### 6. Count the most popular functions from our huge list and take the top 20 for this moment.

In [10]:
cnt = Counter(all_features)
cnt.most_common(21)

[('Elevator', 25398),
 ('HardwoodFloors', 23159),
 ('CatsAllowed', 23148),
 ('DogsAllowed', 21662),
 ('Doorman', 20497),
 ('Dishwasher', 20095),
 ('NoFee', 17806),
 ('LaundryinBuilding', 16093),
 ('FitnessCenter', 13000),
 ('Pre-War', 8978),
 ('LaundryinUnit', 8448),
 ('RoofDeck', 6423),
 ('OutdoorSpace', 5137),
 ('DiningRoom', 4901),
 ('HighSpeedInternet', 4225),
 ('', 3109),
 ('Balcony', 2898),
 ('SwimmingPool', 2648),
 ('LaundryInBuilding', 2565),
 ('NewConstruction', 2507),
 ('Terrace', 2179)]

### 8. Now create 20 new features based on the top 20 values: 1 if the value is in the "Feature" column, otherwise 0.

In [11]:
feature_list = [
    'Elevator',
    'HardwoodFloors',
    'CatsAllowed',
    'DogsAllowed',
    'Doorman',
    'Dishwasher',
    'NoFee',
    'LaundryinBuilding',
    'FitnessCenter',
    'Pre-War',
    'LaundryinUnit',
    'RoofDeck',
    'OutdoorSpace',
    'DiningRoom',
    'HighSpeedInternet',
    'Balcony',
    'SwimmingPool',
    'LaundryInBuilding',
    'NewConstruction',
    'Terrace',
]

for feature in feature_list:
  df_train[feature] = [1 if feature in features else 0 for features in df_train['features']]
  df_test[feature] = [1 if feature in features else 0 for features in df_test['features']]

df_train.head(3)

,bathrooms,bedrooms,building_id,created,description,display_address,features,latitude,listing_id,longitude,...,LaundryinUnit,RoofDeck,OutdoorSpace,DiningRoom,HighSpeedInternet,Balcony,SwimmingPool,LaundryInBuilding,NewConstruction,Terrace
4,1.0,1,8579a0b0d54db803821a35a4a615e97a,2016-06-16 05:55:27,Spacious 1 Bedroom 1 Bathroom in Williamsburg!...,145 Borinquen Place,"DiningRoom,Pre-War,LaundryinBuilding,Dishwashe...",40.7108,7170325,-73.9539,...,0,0,0,1,0,0,0,0,0,0
6,1.0,2,b8e75fc949a6cd8225b455648a951712,2016-06-01 05:44:33,BRAND NEW GUT RENOVATED TRUE 2 BEDROOMFind you...,East 44th,"Doorman,Elevator,LaundryinBuilding,Dishwasher,...",40.7513,7092344,-73.9722,...,0,0,0,0,0,0,0,0,0,0
9,1.0,2,cd759a988b8f23924b5a2058d5ab2b49,2016-06-14 15:19:59,**FLEX 2 BEDROOM WITH FULL PRESSURIZED WALL**L...,East 56th Street,"Doorman,Elevator,LaundryinBuilding,LaundryinUn...",40.7575,7158677,-73.9625,...,1,0,0,0,0,0,0,0,0,0


### 9. Extend our feature set with 'bathrooms', 'bedrooms' and create a special variable feature_list with all feature names. Now we have 22 values. All models should be trained on these 22 features.

In [12]:
feature_list.extend(['bathrooms', 'bedrooms'])
feature_list

['Elevator',
 'HardwoodFloors',
 'CatsAllowed',
 'DogsAllowed',
 'Doorman',
 'Dishwasher',
 'NoFee',
 'LaundryinBuilding',
 'FitnessCenter',
 'Pre-War',
 'LaundryinUnit',
 'RoofDeck',
 'OutdoorSpace',
 'DiningRoom',
 'HighSpeedInternet',
 'Balcony',
 'SwimmingPool',
 'LaundryInBuilding',
 'NewConstruction',
 'Terrace',
 'bathrooms',
 'bedrooms']

## 4. Models implementation — Linear regression

In [13]:
X_train = df_train[feature_list]
y_train = df_train['price']

X_test = df_test[feature_list]
y_test = df_test['price']

### 1. Implement a Python class for a linear regression algorithm with two basic methods — fit and predict. Use stochastic gradient descent (SGD) to find optimal model weights. For better understanding, we recommend implementing separate versions of the algorithm with the analytical solution and non-stochastic gradient descent under the hood.

In [14]:
class LinearRegressionCustom:
  def __init__(self, learning_rate=0.01, n_epochs=1000):
    self.learning_rate = learning_rate
    self.n_epochs = n_epochs
    self.weights = None

  def _add_intercept(self, X):
    """
    Добавляет столбец единиц к матрице X.
    Это нужно, чтобы учесть свободный член (bias) в векторе весов.
    """
    intercept = np.ones((X.shape[0], 1))
    return np.concatenate((intercept, X), axis=1)

  def predict(self, X):
    """
    формула предсказания: y = X * w
    """
    if self.weights is None:
      raise Exception("Model is not trained yet")
    
    print(f"weights={self.weights}")
    X_b = self._add_intercept(X)
    return X_b.dot(self.weights) # скалярное произведение в numpy
  
  # Стохастический градиентный спуск (SGD)
  def fit_sgd(self, X, y):

    if isinstance(X, (np.generic, np.ndarray)):
        X_np = X
    else:
        X_np = X.values # Извлекаем значения из DataFrame
        
    if isinstance(y, (np.generic, np.ndarray)):
        y_np = y
    else:
        y_np = y.values.flatten() # Извлекаем значения из Series и делаем плоским
    
    X_b = self._add_intercept(X_np)
    m = X_np.shape[0]
    n = X_np.shape[1]
    print(f"m={m}, n={n}")

    self.weights = np.random.randn(n + 1) # +1 bias col

    for epoch in range(self.n_epochs):
      # Для SGD важно перемешивать данные каждую эпоху,
      # чтобы циклы обучения не "привыкали" к порядку данных
      indices = np.random.permutation(m)
      X_b_shuffled = X_b[indices]
      y_shuffled = y_np[indices]

      for i in range(m):
        # берем ОДИН случайый пример
        xi = X_b_shuffled[i:i+1]
        yi = y_shuffled[i:i+1]

        # 1. Предсказание для одного примера
        prediction = xi.dot(self.weights)

        # 2. Ошибка для одного примера
        error = prediction - yi

        # 3. Градиент для одного примера
        # (xi.T имеет размерность (n,1), error — скаляр)
        gradient = 2 * xi.T.dot(error)

        # 4. Обновление весов
        # В SGD часто уменьшают learning_rate со временем, 
        # но для простоты оставим его фиксированным
        self.weights = self.weights - self.learning_rate * gradient.flatten()


### 2. What is determenistic model? Make SGD determenistic.

Детерминированность модели - возможность модели при одних и тех же входных данных **всегда** выдавать **один и тот же** результат.
SGD не детерминирован из-за рандомных функций выбора весов и перемешивания данных, необходимо задать random seed, например 21

### 3. Define the R squared (R2) coefficient and implement a function to calculate it.

1. Определние коэффициента
R^2 - Коэффициент **детерминации** (R2, R-squared) — это статистическая мера, которая показывает, **насколько хорошо** регрессионная модель (в нашем случае, линейная регрессия) соответствует данным.

R2 показывает, какая доля дисперсии (изменчивости) зависимой переменной (y) объясняется моделью, по сравнению с тем, если бы мы просто использовали среднее значение y.


**Значения R2:**

  - R2=1.0 (или 100%): Модель идеальна. Все предсказанные точки лежат ровно на линии регрессии. Модель идеально объясняет изменчивость данных.

  - R2=0.0 (или 0%): Модель ничего не объясняет. Ее предсказания не лучше, чем использование среднего значения y.

  - R2<0.0: Модель очень плохая. Она хуже, чем использование среднего значения y. Такое может случиться, если модель сильно переобучена или неправильно применена.

**2. Формула R2**
Для расчета R2 нам нужно сравнить три типа дисперсии (суммы квадратов):

  1. Total Sum of Squares (SStotal​): Общая дисперсия данных y. Считает, насколько фактические значения y отклоняются от их среднего (yˉ​).

    SStotal​=i=1∑n​(yi​−yˉ​)2

  2. Residual Sum of Squares (SSres​): Остаточная дисперсия (ошибка). Считает, насколько фактические значения y отклоняются от предсказаний модели (y^​).
  
    SSres​=i=1∑n​(yi​−y^​i​)2

Формула для R2 сравнивает эти две суммы:
R2= 1 - SSres / SStotal​​

Логика: Если ошибка модели (SSres​) мала по сравнению с общей дисперсией (SStotal​), то R2 будет близок к 1.

In [15]:
def r2_score(y_true, y_pred):
  """
  Рассчитывает коэффициент детерминации (R-squared).
    Args:
        y_true (np.ndarray): Фактические значения целевой переменной.
        y_pred (np.ndarray): Предсказанные моделью значения.
        
    Returns:
        float: Значение R^2.
  """

  # 1. Вычисляем среднее значение y_true (y_bar)
  y_mean = np.mean(y_true)

  # 2. Total Sum of Squares (SS_total) - Сумма квадратов отклонений от среднего
  SS_total = np.sum((y_true - y_mean)**2)

  # 3. Residual Sum of Squares (SS_res) - Сумма квадратов ошибок (остатков)
  SS_res = np.sum((y_true - y_pred)**2)

  # 4. Расчет R^2
  R2 = 1 - (SS_res / SS_total)

  return R2

### 4. Make predictions with your algorithm and estimate the model with MAE, RMSE and R2 metrics.

In [16]:
def mae_score(y_true, y_pred):
  """
  Рассчитывает Среднюю Абсолютную Ошибку (MAE).
  """
  return np.mean(np.abs(y_true - y_pred))

def rmse_score(y_true, y_pred):
  """
  Рассчитывает Корень из Среднеквадратичной Ошибки (RMSE).
  """
  # Сначала считаем MSE (Mean Squared Error)
  mse = np.mean((y_true - y_pred)**2)
  # Затем берем корень
  return np.sqrt(mse)

In [17]:
my_lr = LinearRegressionCustom(n_epochs=100)
my_lr.fit_sgd(X=X_train, y=y_train)

m=48379, n=22


In [18]:
my_predictions_test = my_lr.predict(X_test)

weights=[ 350.59595657  409.85728724 -318.85245832   56.97957911  148.69522374
  526.822648     -7.81507051 -453.61698767 -134.1839309   256.76327235
  -29.54442431  353.19354957   -3.40053427 -104.38853375   86.03055182
  -75.8077341   -21.1001662   311.5892906  -247.4684527  -139.31480094
  134.10262826 1870.59895686  180.71938443]


In [19]:
my_predictions_train = my_lr.predict(X_train)

weights=[ 350.59595657  409.85728724 -318.85245832   56.97957911  148.69522374
  526.822648     -7.81507051 -453.61698767 -134.1839309   256.76327235
  -29.54442431  353.19354957   -3.40053427 -104.38853375   86.03055182
  -75.8077341   -21.1001662   311.5892906  -247.4684527  -139.31480094
  134.10262826 1870.59895686  180.71938443]


In [20]:
def create_metric_df(model_name, y_train_true, y_train_pred, y_test_true, y_test_pred):
    """Создает DataFrame для одной модели, собирая R2, MAE и RMSE."""
    
    # Расчет всех метрик
    metrics = {
        'R2': (r2_score(y_train_true, y_train_pred), r2_score(y_test_true, y_test_pred)),
        'MAE': (mae_score(y_train_true, y_train_pred), mae_score(y_test_true, y_test_pred)),
        'RMSE': (rmse_score(y_train_true, y_train_pred), rmse_score(y_test_true, y_test_pred)),
    }
    
    # Создание DataFrame
    df = pd.DataFrame({
        'Model': model_name,
        'Metric': list(metrics.keys()),
        'Train': [m[0] for m in metrics.values()],
        'Test': [m[1] for m in metrics.values()]
    })
    return df

In [21]:
df_metrics_custom_lr = create_metric_df(
  'LR Custom (SGD)', 
  y_train,
  my_predictions_train,
  y_test,
  my_predictions_test
)

df_metrics_custom_lr

,Model,Metric,Train,Test
0,LR Custom (SGD),R2,0.459712,0.464683
1,LR Custom (SGD),MAE,803.891005,789.425438
2,LR Custom (SGD),RMSE,1174.338918,1141.372992


### 5. Initialize LinearRegression() from sklearn.linear_model, fit the model, and predict the training and test parts as in the previous lesson.

In [22]:
lr = LinearRegression(fit_intercept=True)
lr.fit(X_train, y_train)
predictions_test = lr.predict(X_test)
predictions_train = lr.predict(X_train)

In [23]:
df_metrics_lr = create_metric_df(
  'LR', 
  y_train,
  predictions_train,
  y_test,
  predictions_test
)

df_metrics_lr

,Model,Metric,Train,Test
0,LR,R2,0.580031,0.578628
1,LR,MAE,711.723793,703.312990
2,LR,RMSE,1035.355320,1012.640314


### 6. Compare the quality metrics and make sure the difference is small (between your implementations and sklearn).

### 7. Store the metrics as in the previous lesson in a table with columns model, train, test for MAE table, RMSE table, and R2 coefficient.

In [24]:
metrics_table = pd.concat([df_metrics_custom_lr, df_metrics_lr], ignore_index=True)
metrics_table

,Model,Metric,Train,Test
0,LR Custom (SGD),R2,0.459712,0.464683
1,LR Custom (SGD),MAE,803.891005,789.425438
2,LR Custom (SGD),RMSE,1174.338918,1141.372992
3,LR,R2,0.580031,0.578628
4,LR,MAE,711.723793,703.312990
5,LR,RMSE,1035.355320,1012.640314


## 5. Regularized models implementation — Ridge, Lasso, ElasticNet

### 1. Implement Ridge, Lasso, ElasticNet algorithms: extend the loss function with L2, L1 and both regularizations accordingly.

In [25]:
class LinearRegressionCustom:
  # + параметры регуляризации в __init__
  def __init__(self, learning_rate=0.01, n_epochs=1000, alpha=0.0, l1_ratio=0.0):
    self.learning_rate = learning_rate
    self.n_epochs = n_epochs
    self.weights = None
    # alpha (лямбда) - сила регуляризации
    self.alpha = alpha
    # l1_ratio - соотношения L1/L2 (0.0 - Ridge, 1.0 - Lasso)
    self.l1_ratio = l1_ratio

  def _add_intercept(self, X):
    """
    Добавляет столбец единиц к матрице X.
    Это нужно, чтобы учесть свободный член (bias) в векторе весов.
    """
    intercept = np.ones((X.shape[0], 1))
    return np.concatenate((intercept, X), axis=1)

  def predict(self, X):
    """
    формула предсказания: y = X * w
    """
    if self.weights is None:
      raise Exception("Model is not trained yet")
    
    # print(f"weights={self.weights}")
    X_b = self._add_intercept(X)
    return X_b.dot(self.weights) # скалярное произведение в numpy
  
  # Стохастический градиентный спуск (SGD) + Регуляризация
  def fit_sgd(self, X, y):

    if isinstance(X, (np.generic, np.ndarray)):
        X_np = X
    else:
        X_np = X.values # Извлекаем значения из DataFrame
        
    if isinstance(y, (np.generic, np.ndarray)):
        y_np = y
    else:
        y_np = y.values.flatten() # Извлекаем значения из Series и делаем плоским
    
    X_b = self._add_intercept(X_np)
    m = X_np.shape[0]
    n = X_np.shape[1]

    self.weights = np.random.randn(n + 1) # +1 bias col

    for epoch in range(self.n_epochs):
      # Для SGD важно перемешивать данные каждую эпоху,
      # чтобы циклы обучения не "привыкали" к порядку данных
      indices = np.random.permutation(m)
      X_b_shuffled = X_b[indices]
      y_shuffled = y_np[indices]

      for i in range(m):
        # берем ОДИН случайый пример
        xi = X_b_shuffled[i:i+1]
        yi = y_shuffled[i:i+1]

        # 1. Предсказание для одного примера
        prediction = xi.dot(self.weights)

        # 2. Ошибка для одного примера
        error = prediction - yi

        # 3. Градиент MSE для одного примера
        # (xi.T имеет размерность (n,1), error — скаляр)
        gradient_mse = 2 * xi.T.dot(error).flatten()

        # 4. Градиент регуляризации
        # Создаем вектор штрафа, который не штрафует w0 (bias) - self.weights[0]

        # копируем веса для штрафа, обнуляя w0: [0, w1, w2, ...]
        weights_for_penalty = self.weights.copy()
        weights_for_penalty[0] = 0.0

        # L2 (Ridge) : 2 * alpha * (1 - l1_ratio) * w
        l2_penalty_grad = (2 * self.alpha * (1 - self.l1_ratio)) * weights_for_penalty

        # L1 (Lasso) : alpha * l1_ratio * sign(w)
        l1_penalty_grad = (self.alpha * self.l1_ratio) * np.sign(weights_for_penalty)

        # Полный градиент штрафа
        gradient_penalty = l2_penalty_grad + l1_penalty_grad

        # 5. Обновление весов (штраф добавляется к градиенту)
        full_gradient = gradient_mse + gradient_penalty

    
        # В SGD часто уменьшают learning_rate со временем, 
        # но для простоты оставим его фиксированным
        self.weights = self.weights - self.learning_rate * full_gradient


### 2. Make predictions with your algorithm and estimate the model with MAE, RMSE and R2 metrics.

In [26]:
# 1. Ridge Regression (L2): alpha=0.1, l1_ratio=0.0
model_L1 = LinearRegressionCustom(n_epochs=100, alpha=0.1, l1_ratio=1.0)
# 2. Lasso Regression (L1): alpha=0.1, l1_ratio=1.0
model_L2 = LinearRegressionCustom(n_epochs=100, alpha=0.1, l1_ratio=0.0)
# 3. ElasticNet: alpha=0.1, l1_ratio=0.5 (смесь)
model_elastic = LinearRegressionCustom(n_epochs=100, alpha=0.1, l1_ratio=0.5)

In [27]:
model_L1.fit_sgd(X_train, y_train)
my_y_pred_test = model_L1.predict(X_test)
my_y_pred_train = model_L1.predict(X_train)

df_metrics_L1 = create_metric_df(
  "LR Custom L1",
  y_train,
  my_y_pred_train,
  y_test,
  my_y_pred_test
)

df_metrics_L1

,Model,Metric,Train,Test
0,LR Custom L1,R2,0.558954,0.553218
1,LR Custom L1,MAE,757.836512,751.282547
2,LR Custom L1,RMSE,1061.018257,1042.726128


In [28]:
model_L2.fit_sgd(X_train, y_train)
my_y_pred_test = model_L2.predict(X_test)
my_y_pred_train = model_L2.predict(X_train)

df_metrics_L2 = create_metric_df(
  "LR Custom L2",
  y_train,
  my_y_pred_train,
  y_test,
  my_y_pred_test
)

df_metrics_L2

,Model,Metric,Train,Test
0,LR Custom L2,R2,0.509130,0.496714
1,LR Custom L2,MAE,814.072223,808.563723
2,LR Custom L2,RMSE,1119.345234,1106.698874


In [29]:
model_elastic.fit_sgd(X_train, y_train)
my_y_pred_test = model_elastic.predict(X_test)
my_y_pred_train = model_elastic.predict(X_train)

df_metrics_elastic = create_metric_df(
  "LR Custom Elastic",
  y_train,
  my_y_pred_train,
  y_test,
  my_y_pred_test
)

df_metrics_elastic

,Model,Metric,Train,Test
0,LR Custom Elastic,R2,0.552930,0.546884
1,LR Custom Elastic,MAE,729.748211,723.921293
2,LR Custom Elastic,RMSE,1068.239567,1050.090900


### 3. Initialize Ridge(), Lasso(), and ElasticNet() from sklearn.linear_model, fit the model, and make predictions for the training and test samples as in the previous lesson.

In [30]:
lr_ridge = Ridge(alpha=0.1)
lr_ridge.fit(X_train, y_train)
y_pred_test = lr_ridge.predict(X_test)
y_pred_train = lr_ridge.predict(X_train)

df_metrics_lr_ridge = create_metric_df(
  "LR L2 Ridge",
  y_train,
  y_pred_train,
  y_test, 
  y_pred_test
)

df_metrics_lr_ridge

,Model,Metric,Train,Test
0,LR L2 Ridge,R2,0.580031,0.578628
1,LR L2 Ridge,MAE,711.723466,703.312723
2,LR L2 Ridge,RMSE,1035.355320,1012.640515


In [31]:
lr_lasso = Lasso(alpha=0.1)
lr_lasso.fit(X_train, y_train)
y_pred_test = lr_lasso.predict(X_test)
y_pred_train = lr_lasso.predict(X_train)

df_metrics_lr_lasso = create_metric_df(
  "LR L1 Lasso",
  y_train,
  y_pred_train,
  y_test, 
  y_pred_test
)

df_metrics_lr_lasso

,Model,Metric,Train,Test
0,LR L1 Lasso,R2,0.580029,0.578642
1,LR L1 Lasso,MAE,711.660469,703.245922
2,LR L1 Lasso,RMSE,1035.357221,1012.623256


In [32]:
lr_elastic = ElasticNet(alpha=0.1)
lr_elastic.fit(X_train, y_train)
y_pred_test = lr_elastic.predict(X_test)
y_pred_train = lr_elastic.predict(X_train)

df_metrics_lr_elastic = create_metric_df(
  "LR Elastic",
  y_train,
  y_pred_train,
  y_test, 
  y_pred_test
)

df_metrics_lr_elastic

,Model,Metric,Train,Test
0,LR Elastic,R2,0.568472,0.564291
1,LR Elastic,MAE,716.352336,709.249448
2,LR Elastic,RMSE,1049.506757,1029.722672


### 4. Compare quality metrics and make sure the difference is small (between your implementations and sklearn).

### 5. Store the metrics as in the previous lesson in a table with columns model, train, test for MAE table, RMSE table, and R2 coefficient.

In [33]:
metrics_table = pd.concat([metrics_table, 
                           df_metrics_L1,
                           df_metrics_lr_lasso, 
                           df_metrics_L2, 
                           df_metrics_lr_ridge, 
                           df_metrics_elastic, 
                           df_metrics_lr_elastic], 
                           ignore_index=True)
metrics_table

,Model,Metric,Train,Test
0,LR Custom (SGD),R2,0.459712,0.464683
1,LR Custom (SGD),MAE,803.891005,789.425438
2,LR Custom (SGD),RMSE,1174.338918,1141.372992
3,LR,R2,0.580031,0.578628
4,LR,MAE,711.723793,703.312990
5,LR,RMSE,1035.355320,1012.640314
6,LR Custom L1,R2,0.558954,0.553218
7,LR Custom L1,MAE,757.836512,751.282547
8,LR Custom L1,RMSE,1061.018257,1042.726128
9,LR L1 Lasso,R2,0.580029,0.578642


## 6. Feature normalization

### 1. First, write several examples of why and where feature normalization is mandatory and vice versa.

Нормализация признаков (Feature Normalization) необходима, когда алгоритм **чувствителен к масштабу** входных данных, и она может быть необязательна или нежелательна, когда алгоритм **инвариантен к масштабу**. Нормализация обязательна для алгоритмов, основанных на расстоянии или градиентном спуске, поскольку признаки с большим масштабом могут доминировать над признаками с меньшим масштабом, что приводит к неоптимальной или нестабильной работе.

Примеры, когда нормализация мандаторна

  1. Алгоритмы, основанные на расстоянии (Distance-based Algorithms):
      - **k-ближайших соседей (k-NN) и k-средних (k-Means):** Эти алгоритмы используют Евклидово расстояние. Если один признак измеряется в миллионах (например, "Цена дома"), а другой — в десятках (например, "Количество комнат"), расстояние будет почти полностью определяться признаком с большим масштабом. Нормализация обеспечивает равный вклад всех признаков.

  2. Алгоритмы, основанные на градиентном спуске (Gradient Descent-based Algorithms):

      - **Линейная регрессия с SGD/Регуляризацией (Ridge, Lasso)**: В SGD ненормированные признаки приводят к сильно вытянутому (эллиптическому) ландшафту функции потерь. Градиентный спуск будет "метаться" туда-сюда, делая обучение очень медленным и нестабильным, так как для разных признаков потребуется разный шаг обучения (η).

  3. **Нейронные сети (Neural Networks)**: Нормализация входных данных (обычно в диапазон [0, 1] или со средним 0 и дисперсией 1) помогает стабилизировать активации нейронов и предотвращает проблему исчезающих/взрывающихся градиентов (vanishing/exploding gradients) в начальных слоях.


Примеры, когда нормализация необязательна/не нужна

  1. **Деревья решений (Decision Trees) и Случайный лес (Random Forest):**
      - Эти алгоритмы основаны на пороговых значениях и ранжировании (порядке) признаков, а не на расстоянии. Они выбирают точку разделения, которая максимизирует прирост информации или уменьшает энтропию. Масштаб признака не влияет на то, какое пороговое значение будет выбрано.

  2. **Наивный Байес (Naive Bayes):**
      - Этот алгоритм основан на вероятностях и не чувствителен к масштабу признаков, хотя его вариант для непрерывных данных (Гауссовский Наивный Байес) может работать лучше с нормированными данными.

  3. **Линейная регрессия с Аналитическим решением (OLS):**
      - Если вы используете метод Наименьших Квадратов (формула np.linalg.inv(X.T.dot(X)).dot(X.T).dot(y)), масштабирование не изменит конечных весов (хотя интерпретация весов изменится). Нормализация не нужна для получения корректного решения.

### 2. Let's consider the first of the classical normalization methods — MinMaxScaler. Write a mathematical formula for this method.

Xscaled = (X - Xmin) / (Xmax - Xmin)

### 3. Implement your own function or class for MinMaxScaler feature normalization.

In [34]:
class MyMinMaxScaler:
  """
  Кастомный класс, реализующий Min-Max Feature Scaling.
  Масштабирует данные в диапазон [0, 1].
  """
  def __init__(self):
    # Хранит минимальные значения для каждого признака
    self.min_ = None
    # Хранит размах (max - min) для каждого признака
    self.range_ = None

  def fit(self, X):
    """
    Вычисляет минимальное значение (min) и размах (max - min) 
    для каждого столбца (признака) в тренировочном наборе X.
    
    Args:
        X (np.ndarray): Тренировочный набор данных.
    
    Returns:
        self: Возвращает сам объект.
    """
    # Если X_train имеет форму (m,), преобразуем его в (m, 1) для корректной работы
    if X.ndim == 1:
        X = X.reshape(-1, 1)

    # Вычисляем минимум по каждому столбцу (axis=0)
    self.min_ = np.min(X, axis=0)
    
    # Вычисляем максимум и размах
    max_ = np.max(X, axis=0)
    self.range_ = max_ - self.min_
    
    # Защита от деления на ноль: если range=0 (все значения одинаковы), 
    # range_ должен быть 1, чтобы избежать ошибки.
    self.range_[self.range_ == 0] = 1
    
    return self

  def transform(self, X):
    """
    Масштабирует данные X, используя сохраненные min и range.
    
    Args:
        X (np.ndarray): Набор данных для преобразования (Train или Test).
        
    Returns:
        np.ndarray: Масштабированный набор данных.
    """
    if self.min_ is None or self.range_ is None:
        raise Exception("First call fit() method")
        
    if X.ndim == 1:
        X = X.reshape(-1, 1)
    
    # Применяем формулу: (X - min) / range
    X_scaled = (X - self.min_) / self.range_

    return X_scaled

  def fit_transform(self, X):
    """
    Сначала вызывает fit(X), а затем transform(X).
    """
    self.fit(X)
    return self.transform(X)

### 4. Initialize MinMaxScaler() from sklearn.preprocessing.

### 5. Compare the feature normalization with your own method and with sklearn.

In [35]:
# X_train до нормализации
display(X_train.head(5))

,Elevator,HardwoodFloors,CatsAllowed,DogsAllowed,Doorman,Dishwasher,NoFee,LaundryinBuilding,FitnessCenter,Pre-War,...,OutdoorSpace,DiningRoom,HighSpeedInternet,Balcony,SwimmingPool,LaundryInBuilding,NewConstruction,Terrace,bathrooms,bedrooms
4,0,1,1,1,0,1,0,1,0,1,...,0,1,0,0,0,0,0,0,1.0,1
6,1,1,0,0,1,1,1,1,0,0,...,0,0,0,0,0,0,0,0,1.0,2
9,1,1,0,0,1,1,0,1,0,0,...,0,0,0,0,0,0,0,0,1.0,2
10,0,0,0,0,0,0,0,0,0,0,...,0,0,0,0,0,0,0,0,1.5,3
15,1,0,0,0,1,0,0,1,1,0,...,0,0,0,0,0,0,0,0,1.0,0


In [36]:
my_scaler = MyMinMaxScaler()
sklearn_scaler = MinMaxScaler()

a = my_scaler.fit_transform(X_train)
b = sklearn_scaler.fit_transform(X_train)

display(a)

,Elevator,HardwoodFloors,CatsAllowed,DogsAllowed,Doorman,Dishwasher,NoFee,LaundryinBuilding,FitnessCenter,Pre-War,...,OutdoorSpace,DiningRoom,HighSpeedInternet,Balcony,SwimmingPool,LaundryInBuilding,NewConstruction,Terrace,bathrooms,bedrooms
4,0.0,1.0,1.0,1.0,0.0,1.0,0.0,1.0,0.0,1.0,...,0.0,1.0,0.0,0.0,0.0,0.0,0.0,0.0,0.10,0.125
6,1.0,1.0,0.0,0.0,1.0,1.0,1.0,1.0,0.0,0.0,...,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.10,0.250
9,1.0,1.0,0.0,0.0,1.0,1.0,0.0,1.0,0.0,0.0,...,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.10,0.250
10,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,...,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.15,0.375
15,1.0,0.0,0.0,0.0,1.0,0.0,0.0,1.0,1.0,0.0,...,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.10,0.000
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
124000,1.0,1.0,0.0,0.0,0.0,1.0,0.0,0.0,0.0,0.0,...,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.10,0.375
124002,1.0,0.0,1.0,1.0,1.0,0.0,1.0,0.0,0.0,0.0,...,1.0,0.0,0.0,0.0,0.0,1.0,0.0,0.0,0.10,0.250
124004,1.0,1.0,1.0,1.0,0.0,1.0,1.0,1.0,0.0,1.0,...,0.0,1.0,0.0,0.0,0.0,0.0,0.0,0.0,0.10,0.125
124008,0.0,0.0,0.0,0.0,0.0,1.0,1.0,0.0,0.0,1.0,...,1.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.10,0.250


In [37]:
display(b)

array([[0.   , 1.   , 1.   , ..., 0.   , 0.1  , 0.125],
       [1.   , 1.   , 0.   , ..., 0.   , 0.1  , 0.25 ],
       [1.   , 1.   , 0.   , ..., 0.   , 0.1  , 0.25 ],
       ...,
       [1.   , 1.   , 1.   , ..., 0.   , 0.1  , 0.125],
       [0.   , 0.   , 0.   , ..., 0.   , 0.1  , 0.25 ],
       [1.   , 1.   , 0.   , ..., 0.   , 0.1  , 0.375]], shape=(48379, 22))

### 6. Repeat the steps from b to e for another normalization method StandardScaler.

Формула : z = X - u / q, где u - среднее арифметическое, а q - стандартное отклонение

In [38]:
class MyStandardScaler:
    """
    Кастомный класс, реализующий Standard Scaler (Z-score нормализацию).
    Центрирует данные (среднее = 0) и приводит дисперсию к 1.
    """
    def __init__(self):
        # Хранит среднее значение (mu) для каждого признака
        self.mean_ = None
        # Хранит стандартное отклонение (sigma) для каждого признака
        self.std_ = None
        
    def fit(self, X):
        """
        Вычисляет среднее значение (mean) и стандартное отклонение (std) 
        для каждого столбца (признака) в тренировочном наборе X.
        
        Args:
            X (np.ndarray): Тренировочный набор данных.
        
        Returns:
            self: Возвращает сам объект.
        """
        # Преобразуем 1D массив в 2D, если необходимо
        if X.ndim == 1:
            X = X.reshape(-1, 1)

        # Вычисляем среднее значение по каждому столбцу (axis=0)
        self.mean_ = np.mean(X, axis=0)
        
        # Вычисляем стандартное отклонение по каждому столбцу (axis=0)
        # В NumPy по умолчанию используется выборочное стандартное отклонение (ddof=0 для std)
        self.std_ = np.std(X, axis=0)
        
        # Защита от деления на ноль: если std=0 (все значения одинаковы), 
        # std_ должен быть 1, чтобы избежать ошибки.
        self.std_[self.std_ == 0] = 1
        
        return self

    def transform(self, X):
        """
        Масштабирует данные X, используя сохраненные mean и std.
        
        Args:
            X (np.ndarray): Набор данных для преобразования (Train или Test).
            
        Returns:
            np.ndarray: Стандартизированный набор данных.
        """
        if self.mean_ is None or self.std_ is None:
            raise Exception("Сначала необходимо вызвать метод fit()")
            
        if X.ndim == 1:
            X = X.reshape(-1, 1)
        
        # Применяем формулу: (X - mu) / sigma
        X_scaled = (X - self.mean_) / self.std_
        
        return X_scaled

    def fit_transform(self, X):
        """
        Сначала вызывает fit(X), а затем transform(X).
        """
        self.fit(X)
        return self.transform(X)

## 7. Fit custom and sklearn models with normalized data

In [39]:
scaler_MinMax = MyMinMaxScaler()
X_MinMax_train = scaler_MinMax.fit_transform(X_train)
X_MinMax_test = scaler_MinMax.transform(X_test)

In [40]:
scaler_Standard = MyStandardScaler()
X_Standard_train = scaler_Standard.fit_transform(X_train)
X_Standard_test = scaler_Standard.transform(X_test)

### 1. Fit all models — Linear Regression, Ridge, Lasso, and ElasticNet — with MinMaxScaler.

### 2. Fit all models — Linear Regression, Ridge, Lasso, and ElasticNet — with StandardScaler.

### 3. Add all results to our dataframe with metrics on samples.

In [41]:
# функция для сбора всех метрик
def get_metrics_for_model(model, X_train, y_train, X_test, y_test, model_name, scale_type):
    """Обучает модель, делает предсказания и возвращает список результатов."""
    
    # Обучение
    # Проверяем, если это наш класс, используем fit_sgd
    if hasattr(model, 'fit_sgd'):
        # Используем детерминированность для стабильности
        model.fit_sgd(X_train, y_train) 
    else:
        # Для Sklearn моделей используем стандартный fit
        model.fit(X_train, y_train)

    # Предсказания
    y_train_pred = model.predict(X_train)
    y_test_pred = model.predict(X_test)
    
    # Сбор результатов
    results = [
        # R2
        {'Model': model_name, 'Metric': 'R2', 'Scale': scale_type, 
         'Train': r2_score(y_train, y_train_pred), 'Test': r2_score(y_test, y_test_pred)},
        # MAE
        {'Model': model_name, 'Metric': 'MAE', 'Scale': scale_type, 
         'Train': mae_score(y_train, y_train_pred), 'Test': mae_score(y_test, y_test_pred)},
        # RMSE
        {'Model': model_name, 'Metric': 'RMSE', 'Scale': scale_type, 
         'Train': rmse_score(y_train, y_train_pred), 'Test': rmse_score(y_test, y_test_pred)}
    ]
    return results

# Настройка списка моделей для цикла
models_to_test = [
    # Custom Models (самописный класс)
    (LinearRegressionCustom(alpha=0.0, l1_ratio=0.0, n_epochs=100), 'Custom LR'),
    (LinearRegressionCustom(alpha=0.1, l1_ratio=0.0, n_epochs=100), 'Custom Ridge'),
    (LinearRegressionCustom(alpha=0.1, l1_ratio=1.0, n_epochs=100), 'Custom Lasso'),
    (LinearRegressionCustom(alpha=0.1, l1_ratio=0.5, n_epochs=100), 'Custom ElasticNet'),

    # Sklearn Models (для сравнения)
    (LinearRegression(), 'Sklearn LR'),
    (Ridge(alpha=0.1), 'Sklearn Ridge'),
    (Lasso(alpha=0.1), 'Sklearn Lasso'),
    (ElasticNet(alpha=0.1, l1_ratio=0.5), 'Sklearn ElasticNet'),
]

In [42]:
all_results = []

# Цикл по типам нормализации
# 1. MinMaxScaler
print("--- Обучение на MinMaxScaler ---")
for model_instance, name in models_to_test:
    results = get_metrics_for_model(
        model_instance, 
        X_MinMax_train, y_train, 
        X_MinMax_test, y_test, 
        name, 'MinMax'
    )
    all_results.extend(results)

# 2. StandardScaler
print("--- Обучение на StandardScaler ---")
for model_instance, name in models_to_test:
    # Важно: нужно создавать НОВЫЙ экземпляр модели для каждого обучения!
    # Иначе модель уже обучена на предыдущем типе нормализации.
    if 'Custom' in name:
        new_model = LinearRegressionCustom(alpha=model_instance.alpha, l1_ratio=model_instance.l1_ratio, n_epochs=100)
    elif 'Ridge' in name:
        new_model = Ridge(alpha=model_instance.alpha)
    elif 'Lasso' in name:
        new_model = Lasso(alpha=model_instance.alpha)
    elif 'ElasticNet' in name:
        new_model = ElasticNet(alpha=model_instance.alpha, l1_ratio=model_instance.l1_ratio)
    else:
        new_model = LinearRegression() # LR
        
    results = get_metrics_for_model(
        new_model, 
        X_Standard_train, y_train, 
        X_Standard_test, y_test, 
        name, 'Standard'
    )
    all_results.extend(results)

--- Обучение на MinMaxScaler ---
--- Обучение на StandardScaler ---


In [43]:
# Создание финального DataFrame
df_metrics_final = pd.DataFrame(all_results)
df_metrics_final = df_metrics_final.sort_values(by=['Model', 'Scale', 'Metric']).reset_index(drop=True)
display(df_metrics_final)

,Model,Metric,Scale,Train,Test
0,Custom ElasticNet,MAE,MinMax,913.011003,903.171145
1,Custom ElasticNet,R2,MinMax,0.268442,0.270693
2,Custom ElasticNet,RMSE,MinMax,1366.485671,1332.224826
3,Custom ElasticNet,MAE,Standard,1068.094693,1069.901050
4,Custom ElasticNet,R2,Standard,0.221843,0.189674
5,Custom ElasticNet,RMSE,Standard,1409.335483,1404.274863
6,Custom LR,MAE,MinMax,745.779816,735.333563
7,Custom LR,R2,MinMax,0.525534,0.525629
8,Custom LR,RMSE,MinMax,1100.483410,1074.437362
9,Custom LR,MAE,Standard,946.272769,948.241736


## 8. Overfit models

### 1. Let's look at an overfitted model in practice. From theory, you know that polynomial regression is easy to overfit. So let's create a toy example and see how regularization works in real life.

### 2. In the previous lesson, we created polynomial features with degree 10. Here we repeat these steps from the previous lesson, remembering that we have only 2 basic features — 'bathrooms' and 'bedrooms'.

### 3. And train and fit all our implemented algorithms — Linear Regression, Ridge, Lasso, and ElasticNet — on a set of polynomial features.

### 4. Store the results of the quality metrics in the result dataframe.

In [44]:
def get_metrics_results(model, X_train, y_train, X_test, y_test, model_name):
    """Обучает модель, делает предсказания и возвращает словарь результатов."""
    
    # Обучение
    model.fit_sgd(X_train, y_train) 

    # Предсказания
    y_train_pred = model.predict(X_train)
    y_test_pred = model.predict(X_test)
    
    # Сбор результатов
    results = [
        {'Model': model_name, 'Metric': 'R2', 'Train': r2_score(y_train, y_train_pred), 'Test': r2_score(y_test, y_test_pred)},
        {'Model': model_name, 'Metric': 'MAE', 'Train': mae_score(y_train, y_train_pred), 'Test': mae_score(y_test, y_test_pred)},
        {'Model': model_name, 'Metric': 'RMSE', 'Train': rmse_score(y_train, y_train_pred), 'Test': rmse_score(y_test, y_test_pred)}
    ]
    return results

In [45]:
feature_names = ['bathrooms', 'bedrooms']

X_basic_train = X_train[feature_names]
X_basic_test = X_test[feature_names]

# # 1. Создание полиномиальных признаков
poly_features = PolynomialFeatures(degree=10, include_bias=False) # Не включаем bias, т.к. его добавит LR
X_poly_train = poly_features.fit_transform(X_basic_train)
X_poly_test = poly_features.transform(X_basic_test)

print(f"Количество признаков до полиномиального преобразования: {X_basic_train.shape[1]}")
print(f"Количество признаков после Poly(10): {X_poly_train.shape[1]}") # Должно быть (10+1)*(10+2)/2 - 1 = 65

# 2. Нормализация (Стандартизация)
scaler = StandardScaler()

X_poly_train_scaled = scaler.fit_transform(X_poly_train)
X_poly_test_scaled = scaler.transform(X_poly_test)

print("\nДанные готовы к обучению.")

Количество признаков до полиномиального преобразования: 2
Количество признаков после Poly(10): 65

Данные готовы к обучению.


In [46]:
# 1. Определение моделей с фиксированными параметрами
models_to_test = [
    # Линейная регрессия (Без регуляризации)
    ('Poly LR (a=1.0)', LinearRegressionCustom(alpha=1.0, l1_ratio=0.0, n_epochs=100, learning_rate=0.01)),

    # Ridge (L2)
    ('Poly Ridge (a=1.0)', LinearRegressionCustom(alpha=1.0, l1_ratio=0.0, n_epochs=100, learning_rate=0.01)),
    
    # Lasso (L1)
    ('Poly Lasso (a=1.0)', LinearRegressionCustom(alpha=1.0, l1_ratio=1.0, n_epochs=100, learning_rate=0.01)),
    
    # ElasticNet
    ('Poly ElasticNet (a=1.0)', LinearRegressionCustom(alpha=1.0, l1_ratio=0.5, n_epochs=100, learning_rate=0.01)),
]

all_results_initial = []

# 2. Обучение и сбор метрик
for name, model in models_to_test:
    # Важно: для демонстрации мы используем низкий learning_rate и высокое n_epochs
    print(f"Обучение: {name}...")
    results = get_metrics_results(
        model, 
        X_poly_train_scaled, y_train, 
        X_poly_test_scaled, y_test, 
        name
    )

    all_results_initial.extend(results)

# 3. Создание DataFrame
df_initial_metrics = pd.DataFrame(all_results_initial)

Обучение: Poly LR (a=1.0)...
Обучение: Poly Ridge (a=1.0)...
Обучение: Poly Lasso (a=1.0)...


/tmp/ipykernel_124890/2522207374.py:70: RuntimeWarning: overflow encountered in multiply
  gradient_mse = 2 * xi.T.dot(error).flatten()
/tmp/ipykernel_124890/2522207374.py:80: RuntimeWarning: invalid value encountered in multiply
  l2_penalty_grad = (2 * self.alpha * (1 - self.l1_ratio)) * weights_for_penalty


Обучение: Poly ElasticNet (a=1.0)...


In [47]:
df_initial_metrics

,Model,Metric,Train,Test
0,Poly LR (a=1.0),R2,-1.293197e+00,-3.181344e-01
1,Poly LR (a=1.0),MAE,1.183341e+03,1.141002e+03
2,Poly LR (a=1.0),RMSE,2.419365e+03,1.791027e+03
3,Poly Ridge (a=1.0),R2,-2.484980e+05,-1.788502e+04
4,Poly Ridge (a=1.0),MAE,1.612437e+05,1.364900e+05
5,Poly Ridge (a=1.0),RMSE,7.964217e+05,2.086311e+05
6,Poly Lasso (a=1.0),R2,1.000000e+00,1.000000e+00
7,Poly Lasso (a=1.0),MAE,NaN,NaN
8,Poly Lasso (a=1.0),RMSE,NaN,NaN
9,Poly ElasticNet (a=1.0),R2,-1.254401e+08,-1.786668e+07


### 5. Analyze the results and select the best model according to your opinion.

**Анализ резульатов переобучения**

Проведенный эксперимент на полиномиальных признаках 10-й степени и результаты метрик ярко демонстрируют последствия **переобучения** и **численной нестабильности** в сложных моделях.

  **1. Численная нестабильность (Nan и огромные числа)**
  Результаты Poly Lasso и Poly ElasticNet:
  - Poly Lasso (a=0.1): Дает R2=1.0 и MAE/RMSE=NaN. Это ложный результат, вызванный взрывом градиентов и численной нестабильностью (как обсуждалось ранее). Причина в том, что learning_rate=0.01 слишком высок для 65 признаков, даже с регуляризацией L1.
  - Poly ElasticNet (a=0.1, 0.5): Имеет катастрофически отрицательные R2 (до -390 тысяч) и огромные ошибки (MAE>76 тысяч). Это указывает на то, что модель развалилась в процессе обучения, и ее веса ушли в сторону неконтролируемых значений.

  **2. Анализ переобучения и регуляризации**
  Мы можем сделать осмысленные выводы только по тем моделям, которые не развалились численно: Poly LR и Poly Ridge.
  - Poly LR: Экстремальное переобучение и нестабильность. Отрицательный R2 на Train и почти нулевой на Test.
  - Poly Ridge: Хорошее обобщение. Более высокий R2 на Test, чем на Train, что крайне необычно, но указывает на то, что модель не переобучилась и была стабилизирована L2-штрафом.

**Выбор лучшей модели на основе эксперимента по оверфиту**
По итогам этого прогона, где целью было увидеть, как регуляризация справляется с оверфитом:

Лучшей моделью по критерию стабильности и обобщающей способности (наивысший R2 на тестовом наборе среди стабильных моделей) является Poly Ridge (alpha=1.0).
  - Причина 1: Стабильность: Это единственная модель, которая не развалилась (нет NaN или огромных ошибок) и показала осмысленные метрики.
  - Причина 2: Обобщение: Штраф L2 (Ridge) успешно стабилизировал веса, предотвратил их взрыв и позволил модели показать хоть какой-то положительный R2 на тестовом наборе (0.437), в то время как Poly LR не справилась.

## 9. Native models

### 1. Calculate the mean and median metrics from the previous lesson and add the results to the final dataframe.

In [48]:
# 1. Вычисление наивных предсказаний
# Baseline 1: Среднее значение (mean) тренировочного набора
mean_y_train = np.mean(y_train)

# Предсказание для всех примеров Train и Test - это просто константа!
y_train_pred_mean = np.full(y_train.shape, mean_y_train)
y_test_pred_mean = np.full(y_test.shape, mean_y_train)

# Baseline 2: Медианное значение (median) тренировочного набора
median_y_train = np.median(y_train)
y_train_pred_median = np.full(y_train.shape, median_y_train)
y_test_pred_median = np.full(y_test.shape, median_y_train)

# 2. Функция для сбора метрик Baseline
def get_baseline_metrics(y_train, y_train_pred, y_test, y_test_pred, model_name):
    """Считает метрики для наивного предсказания и возвращает список результатов."""
    return [
        # R2
        {'Model': model_name, 'Metric': 'R2', 'Scale': 'None', 
         'Train': r2_score(y_train, y_train_pred), 'Test': r2_score(y_test, y_test_pred)},
        # MAE
        {'Model': model_name, 'Metric': 'MAE', 'Scale': 'None', 
         'Train': mae_score(y_train, y_train_pred), 'Test': mae_score(y_test, y_test_pred)},
        # RMSE
        {'Model': model_name, 'Metric': 'RMSE', 'Scale': 'None', 
         'Train': rmse_score(y_train, y_train_pred), 'Test': rmse_score(y_test, y_test_pred)}
    ]

# 3. Сбор результатов
baseline_results = []

# Baseline 1: Mean (Среднее)
baseline_results.extend(get_baseline_metrics(
    y_train, y_train_pred_mean, 
    y_test, y_test_pred_mean, 
    'Baseline_Mean'
))

# Baseline 2: Median (Медиана)
baseline_results.extend(get_baseline_metrics(
    y_train, y_train_pred_median, 
    y_test, y_test_pred_median, 
    'Baseline_Median'
))

# 4. Объединение с финальным DataFrame
df_baseline = pd.DataFrame(baseline_results)

# Добавляем наивные результаты к уже собранным результатам
all_results.extend(baseline_results)

# Создание итогового DataFrame (с пересборкой из all_results)
df_metrics_final = pd.DataFrame(all_results)
df_metrics_final = df_metrics_final.sort_values(by=['Model', 'Scale', 'Metric']).reset_index(drop=True)

In [49]:
pd.set_option('display.max_rows', 100)
display(df_metrics_final)

,Model,Metric,Scale,Train,Test
0,Baseline_Mean,MAE,None,1139.192515,1126.239688
1,Baseline_Mean,R2,None,0.000000,-0.000120
2,Baseline_Mean,RMSE,None,1597.646655,1560.084937
3,Baseline_Median,MAE,None,1086.210505,1073.265761
4,Baseline_Median,R2,None,-0.059173,-0.056732
5,Baseline_Median,RMSE,None,1644.236439,1603.632143
6,Custom ElasticNet,MAE,MinMax,913.011003,903.171145
7,Custom ElasticNet,R2,MinMax,0.268442,0.270693
8,Custom ElasticNet,RMSE,MinMax,1366.485671,1332.224826
9,Custom ElasticNet,MAE,Standard,1068.094693,1069.901050


## 10. Compare results

### 1. Print your final tables

### 2. What is the best model?

Лучшей моделью является **Sklearn Lasso** с нормализацией **StandardScaler**.
Sklearn Lasso (строка 46) показала самую высокую точность, достигнув R2=0.578633 на тестовом наборе с минимальной средней абсолютной ошибкой (MAE ≈703).

### 3. Which is the most stable model?

Стабильность модели определяется ее способностью обобщать данные, то есть минимальной разницей между метриками на тренировочном и тестовом наборах.
**Все Sklearn-модели (LR, Lasso, Ridge)** показали очень высокую стабильность с минимальной разницей ΔR2≈0.0014.
Однако, если принять самое минимальное значение разницы как критерий стабильности, то **Custom LR** с нормализацией MinMaxScaler (строка 13) демонстрирует наименьшее расхождение между Train и Test (∣0.525534−0.525629∣≈0.000095).